File: go-backend-interview-prep/06_concurrency/09_worker_pools/main.py
"""Worker pools -- Python equivalent of the Go example.

Two approaches shown:
1. ThreadPoolExecutor (high-level, preferred)
2. Manual threads + queue (matches Go pattern exactly)
"""

import queue
import threading
import time
from concurrent.futures import ThreadPoolExecutor


def process_job(job_id: int, value: int) -> tuple[int, int]:
    """Simulate work: square the value."""
    time.sleep(0.05)
    return job_id, value * value


def main() -> None:
    num_jobs = 9

    # --- Approach 1: ThreadPoolExecutor (idiomatic Python) ---
    print("=== ThreadPoolExecutor ===")
    with ThreadPoolExecutor(max_workers=3) as pool:
        futures = [pool.submit(process_job, j, j) for j in range(1, num_jobs + 1)]

    print("\n=== Results (executor) ===")
    for f in futures:
        job_id, output = f.result()
        print(f"  job {job_id} -> {output}")

    # --- Approach 2: manual threads + queue (matches Go) ---
    print("\n=== Manual threads + queue ===")
    SENTINEL = None
    jobs: queue.Queue = queue.Queue()
    results: queue.Queue = queue.Queue()

    def worker(wid: int):
        while True:
            item = jobs.get()
            if item is SENTINEL:
                break
            jid, val = item
            print(f"  worker {wid}: processing job {jid}")
            time.sleep(0.05)
            results.put((jid, val * val))

    # Start workers
    num_workers = 3
    threads = []
    for w in range(1, num_workers + 1):
        t = threading.Thread(target=worker, args=(w,))
        threads.append(t)
        t.start()

    # Send jobs
    for j in range(1, num_jobs + 1):
        jobs.put((j, j))

    # Send sentinel to each worker
    for _ in range(num_workers):
        jobs.put(SENTINEL)

    for t in threads:
        t.join()

    # Collect results
    print("\n=== Results (manual) ===")
    while not results.empty():
        jid, output = results.get()
        print(f"  job {jid} -> {output}")
    print("all done")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/06_concurrency/09_worker_pools/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// Job represents a unit of work.
type Job struct {
	ID    int
	Value int
}

// Result holds the outcome of processing a Job.
type Result struct {
	JobID  int
	Output int
}

func worker(id int, jobs <-chan Job, results chan<- Result, wg *sync.WaitGroup) {
	defer wg.Done()
	for job := range jobs {
		fmt.Printf("  worker %d: processing job %d\n", id, job.ID)
		time.Sleep(50 * time.Millisecond) // simulate work
		results <- Result{JobID: job.ID, Output: job.Value * job.Value}
	}
}

func main() {
	const numWorkers = 3
	const numJobs = 9

	jobs := make(chan Job, numJobs)
	results := make(chan Result, numJobs)

	// Start workers
	var wg sync.WaitGroup
	for w := 1; w <= numWorkers; w++ {
		wg.Add(1)
		go worker(w, jobs, results, &wg)
	}

	// Send jobs
	for j := 1; j <= numJobs; j++ {
		jobs <- Job{ID: j, Value: j}
	}
	close(jobs) // no more jobs

	// Close results after all workers finish
	go func() {
		wg.Wait()
		close(results)
	}()

	// Collect results
	fmt.Println("\n=== Results ===")
	for r := range results {
		fmt.Printf("  job %d -> %d\n", r.JobID, r.Output)
	}
	fmt.Println("all done")
}
